File: apps/demos/hello-engine/src/recipes/lineRecipe.ts
import type { RecipeBuildResult, RecipeConfig } from "./types";

export function buildLineChartRecipe(cfg: RecipeConfig): RecipeBuildResult {
  // ID allocation: stable + deterministic
  const BUF_LINE = cfg.idBase + 1;
  const GEO_LINE = cfg.idBase + 10;
  const DI_LINE = cfg.idBase + 100;
  const T_VIEW = cfg.idBase + 2000;

  const commands = [
    { cmd: "createBuffer", id: BUF_LINE },
    { cmd: "createGeometry", id: GEO_LINE, vertexBufferId: BUF_LINE, format: "pos2_clip", strideBytes: 8 },
    { cmd: "createDrawItem", id: DI_LINE, geometryId: GEO_LINE, pipeline: "line2d@1" },

    // D1.5 view transform
    { cmd: "createTransform", id: T_VIEW },
    { cmd: "setTransform", id: T_VIEW, tx: 0, ty: 0, sx: 1, sy: 1 },
    { cmd: "attachTransform", targetId: DI_LINE, transformId: T_VIEW }
  ];

  // Dispose in reverse dependency order
  const dispose = [
    { cmd: "delete", kind: "drawItem", id: DI_LINE },
    { cmd: "delete", kind: "geometry", id: GEO_LINE },
    { cmd: "delete", kind: "buffer", id: BUF_LINE },

    { cmd: "delete", kind: "transform", id: T_VIEW }
  ];

  const subscriptions = [
    { kind: "workerStream", stream: "lineSine" as const, bufferId: BUF_LINE }
  ];

  return { commands, dispose, subscriptions };
}

File: core/src/recipe/LineRecipe.cpp
#include "dc/recipe/LineRecipe.hpp"
#include <string>

namespace dc {

LineRecipe::LineRecipe(Id idBase, const LineRecipeConfig& config)
  : Recipe(idBase), config_(config) {}

RecipeBuildResult LineRecipe::build() const {
  RecipeBuildResult result;

  auto idStr = [](Id id) { return std::to_string(id); };

  // Create commands
  result.createCommands.push_back(
    R"({"cmd":"createBuffer","id":)" + idStr(bufferId()) + R"(,"byteLength":0})");

  result.createCommands.push_back(
    R"({"cmd":"createGeometry","id":)" + idStr(geometryId()) +
    R"(,"vertexBufferId":)" + idStr(bufferId()) +
    R"(,"format":"pos2_clip","vertexCount":2})");

  result.createCommands.push_back(
    R"({"cmd":"createDrawItem","id":)" + idStr(drawItemId()) +
    R"(,"layerId":)" + idStr(config_.layerId) +
    R"(,"name":")" + config_.name + R"("})");

  result.createCommands.push_back(
    R"({"cmd":"bindDrawItem","drawItemId":)" + idStr(drawItemId()) +
    R"(,"pipeline":"line2d@1","geometryId":)" + idStr(geometryId()) + "}");

  if (config_.createTransform) {
    result.createCommands.push_back(
      R"({"cmd":"createTransform","id":)" + idStr(transformId()) + "}");
    result.createCommands.push_back(
      R"({"cmd":"attachTransform","drawItemId":)" + idStr(drawItemId()) +
      R"(,"transformId":)" + idStr(transformId()) + "}");
  }

  // Subscriptions
  result.subscriptions.push_back({bufferId(), geometryId(), VertexFormat::Pos2_Clip});

  // Dispose commands (reverse order)
  if (config_.createTransform) {
    result.disposeCommands.push_back(
      R"({"cmd":"delete","id":)" + idStr(transformId()) + "}");
  }
  result.disposeCommands.push_back(
    R"({"cmd":"delete","id":)" + idStr(drawItemId()) + "}");
  result.disposeCommands.push_back(
    R"({"cmd":"delete","id":)" + idStr(geometryId()) + "}");
  result.disposeCommands.push_back(
    R"({"cmd":"delete","id":)" + idStr(bufferId()) + "}");

  return result;
}

} // namespace dc
